### ISSUE TITLE
Strategies persistence and UI

### ISSUE THREAD

**Issue by decebals** [ta4j/ta4j#411]
My question comes somehow in the context of #401. 
I see in some posts (for example #355) that people work to different trading tools based on TA4J (they are 81 projects on GitHub that use TA4J) but I don't find something finished. Maybe the tools/applications are used for internal use (close source), I don't know.
It's clear that an interest for a such tool exists (#347).
My concret question is: `Is there such a tool publicly available?`
Also, myself, I work on a such tool, and now I am in the point where I can load data in my database (command line or via scheduler), I can display data (table and chart) using TimeSeries, I can see my portfolio (connector to broker).
Next step is to create a strategies editor to have a possibility to edit my strategies, to activate or deactivate them.
The second question is: `Did anyone work/use with such an editor?`
You need a modality to persiste the strategies list, together with rules. An idea is to use a readable format (for example to persist each strategy with rules in XML or JSON fromat). Sure, you can code your strategy in java code but I don't want to compile each time the strategy's java class when I want to modify something inside. For starters an XML file is good, but my wish is to edit each strategy visually (a list with rules and some parameter - fields - for each rule).
Have you ever seen such an editor?

**Comment by team172011** [comment]
Hi @decebals ,

I started implementing some kind of charting application where all the indicators are stored in an xml file. That time I was new to JafaFx and have not worked on it for a long time: https://github.com/team172011/TAcharting 
But I am interested in continuing working on this project, but would also need some support. Here the next step would also be to persistent strategies and rules etc and maybe to split the application in charting, modeling and engine part or something similar.

**Comment by nimo23** [comment]
before we can "serialize/deserialize" a strategy, https://github.com/ta4j/ta4j/issues/398 must be done.

**Comment by decebals** [comment]
In my software (Midas - the code name :) I solved (not completed yet) the problem described by me when I created this issue. Take a look at below screenshot

![image](https://user-images.githubusercontent.com/1248311/61080123-05d77280-a42d-11e9-9ea6-1db32b3ac3e9.png)

Each strategy (`StrategyBuilder` implementation) is an extension (I use PF4J [1] as plugin system) - see `@Extension` annotation on class.
So, each strategy can comes from outside of application encapsulated in a plugin. Put the plugin that contains extension(s) in your application and voila you have a new strategy.
In the screen with backtesting you have a combobox with all strategies found in your application by plugin system (PF4J).
If the strategy is modifiable (see `@Modifiable` annotation on class) then that strategy contains some parameters (see `@Param` annotation on field)  then are displayed automatically when the strategy is selected in UI. The user will fill or not (for example `rsiBarCount` has 2 as default value) the value for each parameter and will run the strategy together with a series and parameters values in backtesting.
Everything is generated automatically on the UI part (web application), you need only to supply the `StrategyBuilder` implementation.

For example, see below other strategy implementation created by other person and supplied as a plugin (jar file with only one class)

![image](https://user-images.githubusercontent.com/1248311/61081084-0e30ad00-a42f-11e9-86b6-0b5aead6c4f7.png)

I added the supplied plugin in the plugins directory and voila, I have a new strategy in my application and I can use it without to write a single line of code :smile:.

This is great because my colleague will focus only on writing the strategy and the system created by me will does the rest.

In backstage, each `@Param` annotation will create a `StrategyParameter` metadata, so it's easy to store the strategy's parameters or values (on backtesting runtime).
My intention is to make this project open source, available on GitHub, when it is prepared.

[1] - https://pf4j.org

**Comment by Shudh** [comment]
I was thinking about this issue too. However serialisation/deserialization was just a part of it. I wanted to give myself or someone the ability to implement a strategy without even writing the strategy code. Probably some ui where exisiting indicators are visible and one could connect them using rules at runtime. I was considering evaluating Drools, then reflection...I like your pf4j approach

**Comment by decebals** [comment]
> I wanted to give myself or someone the ability to implement a strategy without even writing the strategy code. Probably some ui where exisiting indicators are visible and one could connect them using rules at runtime.

In the end, I have the same goal but for the moment it's not a priority for me. I want only to close the circle and to have a functional, stable and easy to improve application.
I spent some time on ta4j and I think that I started to understand it. I have some idea to improve the interconnection between (my) application and ta4j (using an annotations processor, ...) but this is another story. In few words, in ta4j is missing a mechanism to generate a metadata/descriptor for each indicator. I created a such mechanism for strategies (see https://github.com/ta4j/ta4j/issues/411#issuecomment-510634048) and for the moment I'm happy with the solution found (it's scalable, easy to use and understand).

**Comment by decebals** [comment]
I wish to show you some screenshots from my application based on ta4j.

- backtesting (see the strategy parameters editor created automatically based on https://github.com/ta4j/ta4j/issues/411#issuecomment-510634048)
![Screenshot from 2019-07-24 23-13-42](https://user-images.githubusercontent.com/1248311/61826304-ce27ec00-ae6a-11e9-921b-6991ad906191.png)
- stocks (Romanian company)
![Screenshot from 2019-07-24 23-15-21](https://user-images.githubusercontent.com/1248311/61826422-0deed380-ae6b-11e9-9e47-8f5e1ca83c77.png)
- historical quotes
![Screenshot from 2019-07-24 23-15-56](https://user-images.githubusercontent.com/1248311/61826491-2b23a200-ae6b-11e9-96f7-e71263681735.png)
- chart
![chart](https://user-images.githubusercontent.com/1248311/61826511-38d92780-ae6b-11e9-9490-8ea82747de31.png)
- news
![Screenshot from 2019-07-24 23-25-12](https://user-images.githubusercontent.com/1248311/61826557-51494200-ae6b-11e9-8873-1a280e276b62.png)
